doc/dev/dialects.rst

```diff
@@ -0,0 +1,417 @@
+
+    #define GET_OP_CLASSES
+    #include "OpenQASM/OpenQASM.cpp.inc"
+
+Besides ensuring the right MLIR headers for our code are included, we mainly need to insert all the
+auto-generated C++ source files, just like we did for the dialect header. The snippet also
+demonstrates that some methods need to be manually implemented, like the dialect initialization
+function ``OpenQASMDialect::initialize()``. Other methods that are typically manually added include
+operation verifiers and operation folding & canonicalization methods.
+
+The accompanying CMake script ``mlir/lib/OpenQASM/CMakeLists.txt`` will generate a build target
+that other Catalyst components can depend on:
+
+.. code-block:: cmake
+
+    add_mlir_library(MLIROpenQASM
+        OpenQASM.cpp
+
+        DEPENDS
+        MLIROpenQASMIncGen
+    )
+
+Note the naming scheme: ``MLIROpenQASM`` is a name of our choice for the dialect build target, while
+``MLIROpenQASMIncGen`` is a target automatically provided by the `add_mlir_dialect` function from the
+provided TableGen file name (``OpenQASM``). The latter represents the generation of C++ files from
+TableGen.
+
+.. warning::
+
+    For any newly added ``CMakeLists.txt``, be sure to add it to its parent CMake file with
+    ``add_subdirectory(<name of new folder>)``. In this case, both ``mlir/include/CMakeLists.txt``
+    and ``mlir/lib/CMakeLists.txt`` will need to be updated with ``add_subdirectory(OpenQASM)``.
+
+
+Using the dialect
+=================
+
+MLIR's standard tool for testing dialects and compiler passes is the ``opt`` tool (inherited from
+LLVM). The tool parses a program in the textual MLIR format, applies arbitrary passes, and prints
+the transformed program back out. Parsing and printing out a program without any transformations
+is also referred to as "round-tripping". Let's see if we can pass this first test with our dialect!
+
+Catalyst comes with its own version of the opt tool, ``quantum-opt``, preloaded with all builtin
+MLIR dialects and transformations, as well all additional compiler components developed for
+Catalyst specifically. Find the file located at ``mlir/tools/quantum-opt/quantum-opt.cpp`` and add
+the following two lines to it:
+
+.. code-block:: c++
+
+    // ...
+    #include "OpenQASM/OpenQASM.h"  // add me
+
+    int main(int argc, char **argv)
+    {
+        // ...
+        registry.insert<catalyst::openqasm::OpenQASMDialect>();  // add me
+
+        // ...
+    }
+
+Similarly, update the corresponding ``mlir/tools/quantum-opt/CMakeLists.txt`` to include the build
+target for our dialect as a dependent library:
+
+.. code-block:: cmake
+
+    # ...
+    set(LIBS
+        # ...
+        MLIROpenQASM  # add me
+    )
+
+    # ...
+
+That's it! We can now build our additions with ``make dialects`` (assuming the project was already
+successfully built once) and test them out.
+
+Save the following test file somewhere and run it through the ``quantum-opt`` tool:
+
+.. code-block:: mlir
```

Comment:
Doesn't look like there are any mlir lexers yet, but I found two gists we could add manually:

- https://gist.github.com/makslevental/1f7f484b5fe09c8cb9c9e4ff7748644a
- https://gist.github.com/oowekyala/110dacc9343dbc1b86b452902d7dc553

Reply:
Do you know to add these? :)

Reply:
Not off the top of my head, but there should be a way of adding them to our `doc/` folder, and registering them with Sphinx.

Reply:
discussed as out of scope